### ISSUE TITLE
...\repositories.yaml: The process cannot access the file...

### ISSUE THREAD

**Issue by clintsinger** [helm/helm#6682]
I am trying Helm 3 Beta 4 and am running into an issue when trying to add a repo.  When I try the same thing with Beta 3 I do not get the error.

**Command**
helm repo add stable https://kubernetes-charts.storage.googleapis.com/

**Result**
Error: read C:\Users\me\AppData\Roaming\helm\repositories.yaml: The process cannot access the file because another process has locked a portion of the file.

Output of `helm version`:
version.BuildInfo{Version:"v3.0.0-beta.4", GitCommit:"7ffc879f137bd3a69eea53349b01f05e3d1d2385", GitTreeState:"dirty", GoVersion:"go1.13.1"}

Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.6", GitCommit:"96fac5cd13a5dc064f7d9f4f23030a6aeface6cc", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:49Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.6", GitCommit:"96fac5cd13a5dc064f7d9f4f23030a6aeface6cc", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:16Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
Docker for Windows

**Comment by XandarZack** [comment]
have the same issue

**Comment by hickeyma** [comment]
@clintsinger Thanks for raising this issue. I want to try and understand better what's happening.
Would you have another process accessing the repo at the same? If you retry a number of times, will it then work?

**Comment by riker09** [comment]
I tried using a powershell with elevated rights (aka Administrator), no success.

I have resetted my Kubernetes Cluster before issuing any commands.

**Comment by riker09** [comment]
@hickeyma The error does not go away when I issue the command multiple times. I have manually deleted the file C:\Users\$USER\AppData\Roaming\helm\repositories.yml (which gets recreated every time the `helm repo add` command is run), but the file is empty.

**Comment by riker09** [comment]
I did a full reset of my Docker on Windows installation and re-installed support for Kubernetes. Still, the error persists.

**Comment by riker09** [comment]
Im trying to work around this issue by downgrading to helm v3.0.0-beta3. I have successfully added the stable repo with `helm repo add stable https://kubernetes-charts.storage.googleapis.com/`. The content of `C:\Users$USER\AppData\Roaming\helm\repositories.yml` is now:

```yaml
apiVersion: ""
generated: "0001-01-01T00:00:00Z"
repositories:
- caFile: ""
  certFile: ""
  keyFile: ""
  name: stable
  password: ""
  url: https://kubernetes-charts.storage.googleapis.com/
  username: ""
```

Switched back to `v3.0.0-beta4` and the command `helm repo list` returns the configured repo. Now I can get back to learning Kubernetes & Helm. 🙂

**Comment by hickeyma** [comment]
@bacongobbler Would you be able to check this on a Windows system?

**Comment by bacongobbler** [comment]
Certainly.

**Comment by ianshore** [comment]
I have the same issue with helm version 2.15.0-rc.1 (client and server versions match).
Windows 10. Other helm repo commands work, like 'helm repo update' & 'helm repo list'.
I've used process explorer to search for any handles to repositories.yaml and nothing's found.
I also recreated the file and able to edit it independently - it's like 'helm repo add' opens it twice internally and conflicts with itself?

**Comment by hickeyma** [comment]
@bacongobbler If you can reproduce the issue, it maybe related to this: https://github.com/gofrs/flock/issues/42#issuecomment-495259713 ?

**Comment by planetf1** [comment]
I can confirm the same issue here - Windows 10 (18999) 
```
version.BuildInfo{Version:"v3.0.0-beta.4", GitCommit:"7ffc879f137bd3a69eea53349b01f05e3d1d2385", GitTreeState:"dirty", GoVersion:"go1.13.1"}
```

**Comment by bacongobbler** [comment]
I can reproduce this issue on Windows. There may be an issue in `flock` that is creating the file but not closing it, causing the dangling file handler (and the eventual error). Taking a closer look at this now.

**Comment by bacongobbler** [comment]
acquiring a lockfile instead of locking the repository file fixes the issue. It is indeed due to https://github.com/gofrs/flock/issues/42 as @hickeyma pointed out:

> The second part is intentional. My goal when building the package was for building lockfiles separate from the data files it was operating on. So you use a common path to a lockfile, and then if you get that exclusive lock you can start working on the data files.

Will submit a PR.

```
PS C:\Users\me\code\go\src\helm.sh\helm> git diff .\cmd\
diff --git a/cmd/helm/repo_add.go b/cmd/helm/repo_add.go
index a28bb82a..bf65fd5e 100644
--- a/cmd/helm/repo_add.go
+++ b/cmd/helm/repo_add.go
@@ -85,8 +85,9 @@ func (o *repoAddOptions) run(out io.Writer) error {
                return err
        }

-       // Lock the repository file for concurrent goroutines or processes synchronization
-       fileLock := flock.New(o.repoFile)
+
+       // Acquire a lock on the repository file for concurrent goroutines or process synchronization
+       fileLock := flock.New(o.repoFile[0:len(o.repoFile)-len(filepath.Ext(o.repoFile))] + "lock")
        lockCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
        defer cancel()
        locked, err := fileLock.TryLockContext(lockCtx, time.Second)
```

**Comment by bacongobbler** [comment]
We'll also need to provide a patch to Helm 2, as this issue is present there as well.

**Comment by bacongobbler** [comment]
fixed in #6692 and #6694.

**Comment by hickeyma** [comment]
Thanks @bacongobbler for testing this on Windows and pushing the fix.

**Comment by alexatnext** [comment]
is this fix released? still experiencing this issue.

**Comment by bacongobbler** [comment]
3.0.0-beta.5 is expected to be released this week (likely tomorrow). You can compile from source in the meantime.

**Comment by celamb4** [comment]
Is there a Helm 2 fix released? Can you specify what are the affected versions so we can downgrade?

**Comment by bacongobbler** [comment]
It was identified in https://github.com/helm/helm/issues/6770 that it was not released in 2.15.0 or 2.15.1. It will be in 2.15.2. Sorry about that!

For now, you may either build from source from the HEAD of the `dev-v2` branch, or downgrade to 2.14.3.

**Comment by intinig** [comment]
2.15.2 doesn't fix this for me.

**Comment by kivi** [comment]
Any workaround until that issue is fixed? I can no install 2.15.2 atm. And in case as the previous comment states: it wasn't fixed or broke again.

**Comment by cvila84** [comment]
according to https://github.com/helm/helm/compare/v2.15.2...dev-v2, the fix was not included in the 2.15.2 release...

**Comment by grolljak** [comment]
Downgrading to 2.14.3 mentioned by @bacongobbler fixed the issue to me.

**Comment by bacongobbler** [comment]
The fix is available in Helm 2.16.0-rc.1.

https://github.com/helm/helm/compare/v2.15.2...v2.16.0-rc.1
